# بسم الله الرحمن الرحيم

In [1]:
from utils import *
import pandas as pd 
from sklearn.model_selection import train_test_split


2023-05-13 09:21:22.076337: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-13 09:21:22.260730: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-05-13 09:21:22.260749: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-05-13 09:21:23.161863: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory
2023-

## Reading the Dataset, cleaning , preparing it for the training 

In [2]:
df = pd.read_csv("dataset/body_level_classification_train.csv")

In [3]:
df_original= cleaning_data(df).copy()

In [4]:
all_class_train,all_class_test=data_spliting(df_original,1/5)

In [5]:
oversampled= over_sampling(all_class_train,sampling_ratio=0.5)

In [6]:
X_train=oversampled.drop(["Body_Level"],axis=1).to_numpy()
Y_train=oversampled["Body_Level"].to_numpy()
X_test =all_class_test.drop(["Body_Level"],axis=1).to_numpy()
Y_test =all_class_test["Body_Level"].to_numpy()

In [7]:
X_train, scaler= data_scaling(X_train)

In [8]:
X_val, X_test, Y_val, Y_test = train_test_split(X_test
                                                     , Y_test, test_size=0.5, random_state=42)

In [9]:
X_val = scaler.transform(X_val)
X_test=scaler.transform(X_test)

In [10]:
model=build_model(2,0.01)

2023-05-13 09:21:25.422386: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2023-05-13 09:21:25.422416: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2023-05-13 09:21:25.422443: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (brownboyComputer): /proc/driver/nvidia/version does not exist
2023-05-13 09:21:25.422753: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [11]:
history = model.fit(X_train,Y_train,validation_data=[X_val,Y_val] 
             ,batch_size=32, epochs=400 , verbose =1, 
                       callbacks=[CustomCallback("NNModels/after_sampling/fifth_model.h5")])

Epoch 1/400
46/46 [==============================] - 1s 6ms/step - loss: 0.9692 - accuracy: 0.4619 - val_loss: 0.4504 - val_accuracy: 0.6463
Epoch 2/400
46/46 [==============================] - 0s 3ms/step - loss: 0.4003 - accuracy: 0.6306 - val_loss: 0.2879 - val_accuracy: 0.6871
Epoch 3/400
46/46 [==============================] - 0s 3ms/step - loss: 0.2973 - accuracy: 0.7095 - val_loss: 0.2449 - val_accuracy: 0.7891
Epoch 4/400
46/46 [==============================] - 0s 2ms/step - loss: 0.2601 - accuracy: 0.7639 - val_loss: 0.2241 - val_accuracy: 0.8299
Epoch 5/400
46/46 [==============================] - 0s 3ms/step - loss: 0.2365 - accuracy: 0.7973 - val_loss: 0.2133 - val_accuracy: 0.8299
Epoch 6/400
46/46 [==============================] - 0s 2ms/step - loss: 0.2185 - accuracy: 0.8163 - val_loss: 0.1948 - val_accuracy: 0.8503
Epoch 7/400
46/46 [==============================] - 0s 2ms/step - loss: 0.2031 - accuracy: 0.8293 - val_loss: 0.1821 - val_accuracy: 0.8571
Epoch 8/400
4

Epoch 57/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0468 - accuracy: 0.9633 - val_loss: 0.1053 - val_accuracy: 0.9796
Epoch 58/400
29/46 [=================>............] - ETA: 0s - loss: 0.0433 - accuracy: 0.9720
Yes You are here 0.9707483053207397 0.9795918464660645
46/46 [==============================] - 0s 3ms/step - loss: 0.0439 - accuracy: 0.9707 - val_loss: 0.1051 - val_accuracy: 0.9796
Epoch 59/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0442 - accuracy: 0.9680 - val_loss: 0.1060 - val_accuracy: 0.9796
Epoch 60/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0434 - accuracy: 0.9667 - val_loss: 0.1050 - val_accuracy: 0.9728
Epoch 61/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0427 - accuracy: 0.9694 - val_loss: 0.1062 - val_accuracy: 0.9796
Epoch 62/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0427 - accuracy: 0.9633 - val_loss: 0.1110 - val_accuracy: 0.9796
Epoch 63/4

Epoch 93/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0281 - accuracy: 0.9802
Yes You are here 0.9768707752227783 0.9795918464660645
46/46 [==============================] - 0s 3ms/step - loss: 0.0292 - accuracy: 0.9769 - val_loss: 0.1111 - val_accuracy: 0.9796
Epoch 94/400
40/46 [=========================>....] - ETA: 0s - loss: 0.0282 - accuracy: 0.9789
Yes You are here 0.9768707752227783 0.9795918464660645
46/46 [==============================] - 0s 3ms/step - loss: 0.0285 - accuracy: 0.9769 - val_loss: 0.1095 - val_accuracy: 0.9796
Epoch 95/400
26/46 [===============>..............] - ETA: 0s - loss: 0.0273 - accuracy: 0.9760
Yes You are here 0.974829912185669 0.9795918464660645
46/46 [==============================] - 0s 3ms/step - loss: 0.0287 - accuracy: 0.9748 - val_loss: 0.1149 - val_accuracy: 0.9796
Epoch 96/400
28/46 [=================>............] - ETA: 0s - loss: 0.0293 - accuracy: 0.9743
Yes You are here 0.9775510430335999 0.9795918464660645
46/46 [===

Epoch 126/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0219 - accuracy: 0.9810 - val_loss: 0.1168 - val_accuracy: 0.9796
Epoch 127/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0217 - accuracy: 0.9769 - val_loss: 0.1144 - val_accuracy: 0.9796
Epoch 128/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0219 - accuracy: 0.9810 - val_loss: 0.1170 - val_accuracy: 0.9796
Epoch 129/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0209 - accuracy: 0.9823 - val_loss: 0.1168 - val_accuracy: 0.9796
Epoch 130/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0219 - accuracy: 0.9803 - val_loss: 0.1208 - val_accuracy: 0.9796
Epoch 131/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0208 - accuracy: 0.9830 - val_loss: 0.1170 - val_accuracy: 0.9796
Epoch 132/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0220 - accuracy: 0.9812
Yes You are here 0.981632649898529 0.98639458

Epoch 174/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0162 - accuracy: 0.9871 - val_loss: 0.1197 - val_accuracy: 0.9796
Epoch 175/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0169 - accuracy: 0.9885
Yes You are here 0.9884353876113892 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0161 - accuracy: 0.9884 - val_loss: 0.1234 - val_accuracy: 0.9864
Epoch 176/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0169 - accuracy: 0.9776 - val_loss: 0.1220 - val_accuracy: 0.9796
Epoch 177/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0160 - accuracy: 0.9864 - val_loss: 0.1213 - val_accuracy: 0.9864
Epoch 178/400
28/46 [=================>............] - ETA: 0s - loss: 0.0151 - accuracy: 0.9888
Yes You are here 0.9850339889526367 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0161 - accuracy: 0.9850 - val_loss: 0.1195 - val_accuracy: 0.9864
Epoch 179

Epoch 213/400
28/46 [=================>............] - ETA: 0s - loss: 0.0150 - accuracy: 0.9855
Yes You are here 0.9870748519897461 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0138 - accuracy: 0.9871 - val_loss: 0.1169 - val_accuracy: 0.9864
Epoch 214/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0147 - accuracy: 0.9857 - val_loss: 0.1182 - val_accuracy: 0.9796
Epoch 215/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0143 - accuracy: 0.9850 - val_loss: 0.1157 - val_accuracy: 0.9796
Epoch 216/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0147 - accuracy: 0.9871 - val_loss: 0.1218 - val_accuracy: 0.9796
Epoch 217/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0144 - accuracy: 0.9871 - val_loss: 0.1204 - val_accuracy: 0.9796
Epoch 218/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0127 - accuracy: 0.9885
Yes You are here 0.9870748519897461 0.986394584178

Epoch 249/400
28/46 [=================>............] - ETA: 0s - loss: 0.0132 - accuracy: 0.9866
Yes You are here 0.9877551198005676 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0126 - accuracy: 0.9878 - val_loss: 0.1180 - val_accuracy: 0.9864
Epoch 250/400
29/46 [=================>............] - ETA: 0s - loss: 0.0113 - accuracy: 0.9903
Yes You are here 0.9891156554222107 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0121 - accuracy: 0.9891 - val_loss: 0.1172 - val_accuracy: 0.9864
Epoch 251/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0125 - accuracy: 0.9857 - val_loss: 0.1215 - val_accuracy: 0.9796
Epoch 252/400
29/46 [=================>............] - ETA: 0s - loss: 0.0093 - accuracy: 0.9957
Yes You are here 0.9897959232330322 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0119 - accuracy: 0.9898 - val_loss: 0.1163 - val_accuracy: 0.9864
Epoch 253/400


Epoch 284/400
33/46 [====================>.........] - ETA: 0s - loss: 0.0106 - accuracy: 0.9877
Yes You are here 0.9877551198005676 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0111 - accuracy: 0.9878 - val_loss: 0.1143 - val_accuracy: 0.9864
Epoch 285/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0112 - accuracy: 0.9864 - val_loss: 0.1136 - val_accuracy: 0.9864
Epoch 286/400
29/46 [=================>............] - ETA: 0s - loss: 0.0115 - accuracy: 0.9914
Yes You are here 0.9918367266654968 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0111 - accuracy: 0.9918 - val_loss: 0.1131 - val_accuracy: 0.9864
Epoch 287/400
28/46 [=================>............] - ETA: 0s - loss: 0.0111 - accuracy: 0.9888
Yes You are here 0.9884353876113892 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0111 - accuracy: 0.9884 - val_loss: 0.1134 - val_accuracy: 0.9864
Epoch 288/400


Epoch 346/400
29/46 [=================>............] - ETA: 0s - loss: 0.0084 - accuracy: 0.9935
Yes You are here 0.9891156554222107 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0103 - accuracy: 0.9891 - val_loss: 0.1097 - val_accuracy: 0.9864
Epoch 347/400
28/46 [=================>............] - ETA: 0s - loss: 0.0092 - accuracy: 0.9955
Yes You are here 0.9931972622871399 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0095 - accuracy: 0.9932 - val_loss: 0.1122 - val_accuracy: 0.9864
Epoch 348/400
39/46 [========================>.....] - ETA: 0s - loss: 0.0092 - accuracy: 0.9904
Yes You are here 0.9904761910438538 0.9863945841789246
46/46 [==============================] - 0s 2ms/step - loss: 0.0097 - accuracy: 0.9905 - val_loss: 0.1088 - val_accuracy: 0.9864
Epoch 349/400
39/46 [========================>.....] - ETA: 0s - loss: 0.0089 - accuracy: 0.9928
Yes You are here 0.9918367266654968 0.9863945841789246
46/46

Epoch 379/400
42/46 [==========================>...] - ETA: 0s - loss: 0.0089 - accuracy: 0.9926
Yes You are here 0.9911564588546753 0.9863945841789246
46/46 [==============================] - 0s 2ms/step - loss: 0.0092 - accuracy: 0.9912 - val_loss: 0.1075 - val_accuracy: 0.9864
Epoch 380/400
34/46 [=====================>........] - ETA: 0s - loss: 0.0085 - accuracy: 0.9936
Yes You are here 0.9931972622871399 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0089 - accuracy: 0.9932 - val_loss: 0.1048 - val_accuracy: 0.9864
Epoch 381/400
24/46 [==============>...............] - ETA: 0s - loss: 0.0083 - accuracy: 0.9948
Yes You are here 0.9938775300979614 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0090 - accuracy: 0.9939 - val_loss: 0.1083 - val_accuracy: 0.9864
Epoch 382/400
29/46 [=================>............] - ETA: 0s - loss: 0.0076 - accuracy: 0.9968
Yes You are here 0.9931972622871399 0.9863945841789246
46/46

In [12]:
accurate_train_accuracy,classify_report= test_model("NNModels/after_sampling/fifth_model.h5",X_train,Y_train)

46/46 [==============================] - 0s 960us/step


In [13]:
accurate_train_accuracy

0.9965986394557823

In [14]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       272
           1       0.99      0.99      0.99       278
           2       0.99      0.99      0.99       376
           3       1.00      1.00      1.00       544

    accuracy                           1.00      1470
   macro avg       1.00      1.00      1.00      1470
weighted avg       1.00      1.00      1.00      1470



In [15]:
accurate_val_accuracy,classify_report= test_model("NNModels/after_sampling/fifth_model.h5",X_val,Y_val)

5/5 [==============================] - 0s 2ms/step


In [16]:
accurate_val_accuracy

0.9863945578231292

In [17]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        20
           1       0.95      1.00      0.98        20
           2       0.97      1.00      0.99        36
           3       1.00      0.99      0.99        71

    accuracy                           0.99       147
   macro avg       0.98      0.98      0.98       147
weighted avg       0.99      0.99      0.99       147



In [18]:
accurate_test_accuracy,classify_report= test_model("NNModels/after_sampling/fifth_model.h5",X_test,Y_test)

5/5 [==============================] - 0s 2ms/step


In [19]:
accurate_test_accuracy

0.9864864864864865

In [20]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        18
           1       1.00      0.95      0.97        20
           2       0.98      0.98      0.98        45
           3       0.98      1.00      0.99        65

    accuracy                           0.99       148
   macro avg       0.99      0.98      0.99       148
weighted avg       0.99      0.99      0.99       148



## Notes

- name            : first model 
- split           : 1/3 
- oversampling    : 0.5 
- val_acc         : 0.9836734693877551
- test_acc        : 0.983739837398374
- gamma           : 1 
---------------------------------------------------------------------------------------------
- name            : second model 
- split           : 1/3 
- oversampling    : 0.8
- val_acc         : 0.9877551020408163
- test_acc        : 0.9878048780487805
- gamma           : 1 

---------------------------------------------------------------------------------------------

- name            : forth model 
- split           : 1/5
- oversampling    : 1
- val_acc         : 0.9863945578231292
- test_acc        : 0.9932432432432432
- gamma           : 2

---------------------------------------------------------------------------------------------
- name            : fifth model 
- split           : 1/5
- oversampling    : 0.5
- val_acc         : 0.9863945578231292
- test_acc        : 0.9864864864864865
- gamma           : 2

---------------------------------------------------------------------------------------------

